statistics/statistics.go

```diff
@@ -212,94 +249,150 @@ func (c *Column) mergeBuckets(bucketIdx int64) {
+// Table represents statistics for a table.
+type Table struct {
+	Info    *model.TableInfo
+	Columns []*Column
+	Indices []*Column
+	Count   int64 // Total row count in a table.
+	Pseudo  bool
+}
+
+// SaveToStorage saves stats table to storage.
+func (t *Table) SaveToStorage(ctx context.Context) error {
+	_, err := ctx.(sqlexec.SQLExecutor).Execute("begin")
+	txn := ctx.Txn()
+	version := txn.StartTS()
+	SetStatisticsTableCache(t.Info.ID, t, version)
+	deleteSQL := fmt.Sprintf("delete from mysql.stats_meta where table_id = %d", t.Info.ID)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(deleteSQL)
 	if err != nil {
-		return nil, errors.Trace(err)
+		return errors.Trace(err)
+	}
+	insertSQL := fmt.Sprintf("insert into mysql.stats_meta (version, table_id, count) values (%d, %d, %d)", version, t.Info.ID, t.Count)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(insertSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteSQL = fmt.Sprintf("delete from mysql.stats_columns where table_id = %d", t.Info.ID)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(deleteSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	deleteSQL = fmt.Sprintf("delete from mysql.stats_buckets where table_id = %d", t.Info.ID)
+	_, err = ctx.(sqlexec.SQLExecutor).Execute(deleteSQL)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, col := range t.Columns {
+		col.saveToStorage(ctx, t.Info.ID, false)
 	}
-	cpb := &ColumnPB{
-		Id:      proto.Int64(col.ID),
-		Ndv:     proto.Int64(col.NDV),
-		Numbers: col.Numbers,
-		Value:   data,
-		Repeats: col.Repeats,
+	for _, idx := range t.Indices {
+		idx.saveToStorage(ctx, t.Info.ID, true)
 	}
-	return cpb, nil
+	_, err = ctx.(sqlexec.SQLExecutor).Execute("commit")
+	return errors.Trace(err)
 }
 
-// columnFromPB gets Column from ColumnPB.
-func columnFromPB(cpb *ColumnPB, ft *types.FieldType) (*Column, error) {
-	var values []types.Datum
-	var err error
-	if len(cpb.GetValue()) > 0 {
-		values, err = codec.Decode(cpb.GetValue(), 1)
-		if err != nil {
-			return nil, errors.Trace(err)
-		}
+func colStatsFromStorage(ctx context.Context, tableID int64, colID int64, tp *types.FieldType, distinct int64, isIndex bool) (*Column, error) {
+	var colName string
+	if isIndex {
+		colName = "index_id"
+	} else {
+		colName = "col_id"
 	}
-	c := &Column{
-		ID:      cpb.GetId(),
-		NDV:     cpb.GetNdv(),
-		Numbers: cpb.GetNumbers(),
-		Values:  make([]types.Datum, len(values)),
-		Repeats: cpb.GetRepeats(),
+	selSQL := fmt.Sprintf("select bucket_id, count, repeats, value from mysql.stats_buckets where table_id = %d and %s = %d", tableID, colName, colID)
+	rows, _, err := ctx.(sqlexec.RestrictedSQLExecutor).ExecRestrictedSQL(ctx, selSQL)
+	if err != nil {
+		return nil, errors.Trace(err)
 	}
-	for i, val := range values {
-		c.Values[i], err = tablecodec.Unflatten(val, ft, false)
-		if err != nil {
-			return nil, errors.Trace(err)
+	bucketSize := len(rows)
+	if bucketSize == 0 {
+		return nil, nil
```

Comment:
Why return nil? Even if it is an empty table, currently it will have at least 1 bucket. `bucketSize` is 0 may mean that this row is added after analyze.